File: _20_valid_parentheses/main.cpp
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, char> matchingParentheses = {
            {'(', ')'},
            {'{', '}'},
            {'[', ']'}
        };

        stack<char> stack;

        for (int i = 0; i < s.size(); i++) {
            if (matchingParentheses.find(s.at(i)) != matchingParentheses.end()) {
                stack.push(s.at(i));
            } else if (!stack.empty() && matchingParentheses[stack.top()] == s.at(i)) {
                stack.pop();
            } else {
                return false;
            }
        }

        return stack.empty();
    }
};

File: _20_valid_parentheses/main.py
class Solution:
    def isValid(self, s: str) -> bool:
        matching = {
            '(': ')',
            '{': '}',
            '[': ']'
        }

        stack = []

        for c in s:
            if c in matching:
                stack.append(matching[c])
            elif len(stack) > 0 and stack[-1] == c:
                stack.pop()
            else:
                return False

        return len(stack) == 0